orthanc/orthanc-anon/plugin/pixl.py

```diff
@@ -219,42 +216,151 @@ def OnHeartBeat(output, uri, **request) -> Any:  # noqa: ARG001
-    elif origin == orthanc.InstanceOrigin.DICOM_PROTOCOL:
-        orthanc.LogWarning("DICOM instance received from the DICOM protocol")
-
-    # It's important that as much code in this handler as possible is inside this "try" block.
-    # This ensures we discard the image if anything goes wrong in the anonymisation process.
-    # If the handler raises an exception the pre-anon image will be kept.
-    try:
-        return _process_dicom_instance(receivedDicom)
-    except Exception:  # noqa: BLE001
-        orthanc.LogError("Failed to anonymize instance due to\n" + traceback.format_exc())
-        return orthanc.ReceivedInstanceAction.DISCARD, None
-
-
-def _process_dicom_instance(receivedDicom: bytes) -> tuple[orthanc.ReceivedInstanceAction, None]:
-    # Read the bytes as DICOM/
-    dataset = dcmread(BytesIO(receivedDicom))
+def _anonymise_dicom_instance(dataset: pydicom.Dataset) -> bytes:
+    """
+    Anonymise a DICOM instance.
 
-    # Attempt to anonymise and drop the study if any exceptions occur
+    Skip the instance for a study if a PIXLDiscardError is raised.
+    """
     try:
         study_identifiers = get_study_info(dataset)
         anonymise_and_validate_dicom(dataset, config_path=None, synchronise_pixl_db=True)
-        return orthanc.ReceivedInstanceAction.MODIFY, write_dataset_to_bytes(dataset)
+        return write_dataset_to_bytes(dataset)
     except PixlDiscardError as error:
-        logger.warning("Skipping {}: {}", study_identifiers, error)
-        return orthanc.ReceivedInstanceAction.DISCARD, None
-    except:  # noqa: E722 Allow bare except
-        # Called from a callback in orthanc so will never take down the service
-        # we want to do dicard anything on failure so we don't leak identifiable information
-        logger.exception("{} failed", study_identifiers)
-        return orthanc.ReceivedInstanceAction.DISCARD, None
+        logger.warning("Skipping instance for study {}: {}", study_identifiers, error)
+        raise
+
+
+def ImportStudyFromRaw(output, uri, **request):  # noqa: ARG001
+    """
+    Import a study from Orthanc Raw.
+
+    - Pull a study from Orthanc Raw based on its resource ID. Wait for the study to be stable.
+    - Iterate over instances and anonymise them
+    - Re-upload the study via the dicom-web api. Wait for the study to be stable.
+    - Notify the PIXL export-api to send the study the to relevant endpoint
+    """
+    payload = json.loads(request["body"])
+    study_uid = payload["StudyInstanceUID"]
+    query_id = payload["QueryID"]
+    retrieve_response = json.loads(
+        orthanc.RestApiPost(f"/queries/{query_id}/retrieve", json.dumps({}))
+    )
+    logger.debug(
+        "Response from retrieving study {} from Orthan Raw: {}", study_uid, retrieve_response
+    )
+
+    # Download the zipped study from Orthanc Anon
+    study_resource_id = _get_existing_study(study_uid=study_uid)
+    if study_resource_id is None:
+        return
```

Comment:
maybe should log this as an warning/error because that's surprising

Reply:
yeah definitely, an exception is now raiesd in `_get_existing_study` (renamed to `_get_study_resource_id`